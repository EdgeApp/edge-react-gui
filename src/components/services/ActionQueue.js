// @flow

import { type EdgeAccount } from 'edge-core-js'

import { updateActionProgramState } from '../../controllers/action-queue/redux/actions'
import { executeActionProgram } from '../../controllers/action-queue/runtime'
import { type ActionQueueMap } from '../../controllers/action-queue/types'
import { useAsyncEffect } from '../../hooks/useAsyncEffect'
import { useRef } from '../../types/reactHooks'
import { useDispatch, useSelector } from '../../types/reactRedux'

export const ActionQueue = () => {
  const dispatch = useDispatch()
  const account: EdgeAccount = useSelector(state => state.core.account)
  const queue: ActionQueueMap = useSelector(state => state.actionQueue.queue)
  const executingRef = useRef<{ [programId: string]: boolean }>({})

  // TODO: Fix sync server stale ActionQueue state overwrite on startup
  //
  // Initialization
  //

  // useAsyncEffect(async () => {
  //   if (account?.dataStore != null) {
  //     const store = makeActionQueueStore(account)
  //     const queue = await store.getActionQueueMap()
  //     dispatch({
  //       type: 'ACTION_QUEUE/LOAD_QUEUE',
  //       data: queue
  //     })
  //   }
  // }, [account, dispatch])

  //
  // Runtime
  //

  useAsyncEffect(async () => {
    if (queue != null) {
      const executing = executingRef.current
      const promises = Object.keys(queue)
        .filter(
          programId =>
            // Ignore running programs
            !executing[programId]
        )
        .map(async programId => {
          // Set program to running
          executing[programId] = true

          const { program, state } = queue[programId]
          const { nextState } = await executeActionProgram(account, program, state).catch((error: Error) => ({
            nextState: {
              ...state,
              effect: { type: 'done', error }
            }
          }))

          // Update program state
          dispatch(updateActionProgramState(nextState))

          // Unset program to running
          executing[programId] = false
        })

      await Promise.all(promises)
    }
  }, [queue])

  return null
}
