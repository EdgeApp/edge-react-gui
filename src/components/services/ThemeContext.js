// @flow

import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { womanTheme } from '../../theme/variables/womanTheme'
import { type Theme } from '../../types/Theme.js'

export type { Theme }
export { cacheStyles }

/**
 * Utility type for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(womanTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
