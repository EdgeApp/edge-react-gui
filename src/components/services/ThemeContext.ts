import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { config } from '../../theme/appConfig'
import { Theme } from '../../types/Theme'

export type { Theme }
export { cacheStyles }

/**
 * Utility for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(config.darkTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
