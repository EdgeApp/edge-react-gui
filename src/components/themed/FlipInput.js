// @flow

import Clipboard from '@react-native-community/clipboard'
import { bns } from 'biggystring'
import * as React from 'react'
import { type Event, Animated, Image, Platform, TextInput, TouchableOpacity, TouchableWithoutFeedback, View } from 'react-native'
import Menu, { MenuOption, MenuOptions, MenuTrigger, renderers } from 'react-native-popup-menu'

import { Fontello } from '../../assets/vector'
import { formatNumberInput, prettifyNumber, truncateDecimals, truncateDecimalsPeriod } from '../../locales/intl.js'
import s from '../../locales/strings.js'
import * as UTILS from '../../util/utils.js'
import { showError } from '../services/AirshipInstance.js'
import { type Theme, type ThemeProps, cacheStyles, withTheme } from '../services/ThemeContext.js'
import { BlinkingCursor } from './BlinkingCursor'
import { EdgeText } from './EdgeText.js'
import { ButtonBox, RightChevronButton } from './ThemedButtons.js'

export type FlipInputFieldInfo = {
  currencyName: string,
  currencySymbol: string, // currency symbol of field
  currencyCode: string, // 3-5 digit currency code

  // Maximum number of decimals to allow the user to enter. FlipInput will automatically truncate use input to this
  // number of decimals as the user types.
  maxEntryDecimals: number,

  // Maximum number of decimals to convert from the opposite field to this field.
  // ie If the user is typing into the fiat field, and this FlipInputFieldInfo refers to a BTC field, then this is the number of
  // decimals to use when converting the fiat value into this crypto field.
  maxConversionDecimals: number
}

type State = {
  isToggled: boolean,
  textInputFrontFocus: boolean,
  textInputBackFocus: boolean,
  overridePrimaryDecimalAmount: string,
  forceUpdateGuiCounter: number,
  primaryDisplayAmount: string, // Actual display amount including 1000s separator and localized for region
  secondaryDisplayAmount: string, // Actual display amount including 1000s separator and localized for region
  primaryDecimalAmount: string,
  secondaryDecimalAmount: string,
  rerenderCounter: number
}

export type FlipInputOwnProps = {
  // Override value of the primary field. This will be the initial value of the primary field. Only changes to this value will
  // cause changes to the primary field
  overridePrimaryDecimalAmount: string,

  // Exchange rate
  exchangeSecondaryToPrimaryRatio: string,

  // Information regarding the primary and secondary field. Mostly related to currency name, code, and denominations
  primaryInfo: FlipInputFieldInfo,
  secondaryInfo: FlipInputFieldInfo,
  onNext?: () => void,
  onFocus?: () => void,
  onBlur?: () => void,
  forceUpdateGuiCounter: number,

  // Callback when primaryDecimalAmount changes. **This is only called when the user types into a field or if
  // exchangeSecondaryToPrimaryRatio changes. This does NOT get called when overridePrimaryDecimalAmount is changed by the parent
  onAmountChanged(decimalAmount: string): void,
  isEditable: boolean,
  isFiatOnTop: boolean,
  isFocus: boolean,

  topReturnKeyType?: string,
  inputAccessoryViewID?: string,
  headerText: string,
  headerLogo: string | void,
  headerCallback?: () => void,
  keyboardVisible: boolean,
  flipInputRef: (FlipInput: any) => void
}

type Props = FlipInputOwnProps & ThemeProps

type Amounts = {
  primaryDecimalAmount: string,
  primaryDisplayAmount: string,
  secondaryDecimalAmount: string,
  secondaryDisplayAmount: string
}

export const sanitizeDecimalAmount = (amount: string, maxEntryDecimals: number): string => {
  // Replace all commas into periods
  amount = amount.replace(',', '.')

  // Remove characters except numbers and decimal separator
  amount = amount.replace(/[^0-9.]/g, '')

  // Trunctuate decimals to limited decimal entries, also remove additional periods
  return truncateDecimalsPeriod(amount, maxEntryDecimals)
}

const checkKeyPress = (keyPressed: string, decimalAmount: string) => {
  if (keyPressed === '.' && decimalAmount.includes('.')) {
    return false
  }

  if (keyPressed.match(/[^0-9.]/)) {
    return false
  }

  return true
}

// Assumes a US locale decimal input
const setPrimaryToSecondary = (props: Props, primaryDecimalAmount: string): Amounts => {
  // Formats into locale specific format. Add currency symbol
  const primaryDisplayAmount = formatNumberInput(prettifyNumber(primaryDecimalAmount))

  // Converts to secondary value using exchange rate
  let secondaryDecimalAmount = bns.mul(primaryDecimalAmount, props.exchangeSecondaryToPrimaryRatio)

  // Truncate to however many decimals the secondary format should have
  secondaryDecimalAmount = UTILS.truncateDecimals(secondaryDecimalAmount, props.secondaryInfo.maxConversionDecimals)

  // Format into locale specific format. Add currency symbol
  const secondaryDisplayAmount = formatNumberInput(prettifyNumber(secondaryDecimalAmount))

  // Set the state for display in render()
  return { primaryDecimalAmount, primaryDisplayAmount, secondaryDecimalAmount, secondaryDisplayAmount }
}

// Pretty much the same as setPrimaryToSecondary
const setSecondaryToPrimary = (props: Props, secondaryDecimalAmount: string): Amounts => {
  const secondaryDisplayAmount = formatNumberInput(prettifyNumber(secondaryDecimalAmount))
  let primaryDecimalAmount = props.exchangeSecondaryToPrimaryRatio === '0' ? '0' : bns.div(secondaryDecimalAmount, props.exchangeSecondaryToPrimaryRatio, 18)
  primaryDecimalAmount = UTILS.truncateDecimals(primaryDecimalAmount, props.primaryInfo.maxConversionDecimals)
  const primaryDisplayAmount = formatNumberInput(prettifyNumber(primaryDecimalAmount))
  return { primaryDisplayAmount, primaryDecimalAmount, secondaryDisplayAmount, secondaryDecimalAmount }
}

const getInitialState = (props: Props) => {
  const state: State = {
    isToggled: false,
    textInputFrontFocus: false,
    textInputBackFocus: false,
    overridePrimaryDecimalAmount: '',
    primaryDisplayAmount: '',
    forceUpdateGuiCounter: 0,
    secondaryDisplayAmount: '',
    primaryDecimalAmount: '',
    secondaryDecimalAmount: '',
    rerenderCounter: 0
  }

  if (props.overridePrimaryDecimalAmount === '') return state

  const primaryDecimalAmount = sanitizeDecimalAmount(props.overridePrimaryDecimalAmount, props.primaryInfo.maxEntryDecimals)
  return Object.assign(state, setPrimaryToSecondary(props, primaryDecimalAmount))
}

class FlipInputComponent extends React.PureComponent<Props, State> {
  animatedValue: Animated.Value
  frontInterpolate: Animated.Value
  backInterpolate: Animated.Value
  androidFrontOpacityInterpolate: Animated.Value
  androidBackOpacityInterpolate: Animated.Value
  textInputFront: TextInput | null
  textInputBack: TextInput | null
  clipboardMenu: any

  constructor(props: Props) {
    super(props)
    this.state = getInitialState(props)

    // Mounting Animation
    this.animatedValue = new Animated.Value(0)
    this.frontInterpolate = this.animatedValue.interpolate({
      inputRange: [0, 1],
      outputRange: ['0deg', '180deg']
    })

    this.backInterpolate = this.animatedValue.interpolate({
      inputRange: [0, 1],
      outputRange: ['180deg', '360deg']
    })
    this.androidFrontOpacityInterpolate = this.animatedValue.interpolate({
      inputRange: [0, 0.5, 0.5],
      outputRange: [1, 1, 0]
    })
    this.androidBackOpacityInterpolate = this.animatedValue.interpolate({
      inputRange: [0.5, 0.5, 1],
      outputRange: [0, 1, 1]
    })
  }

  componentDidMount() {
    this.props.flipInputRef(this)
    setTimeout(() => {
      if (this.props.keyboardVisible && this.props.overridePrimaryDecimalAmount === '0' && this.textInputFront) {
        this.textInputFront.focus()
      }
    }, 400)

    if (this.props.isFiatOnTop) {
      this.setState({
        isToggled: !this.state.isToggled
      })
      Animated.timing(this.animatedValue, {
        toValue: 1,
        duration: 0,
        useNativeDriver: true
      }).start()
      setTimeout(() => {
        this.setState({
          secondaryDisplayAmount: ''
        })
      }, 10)
    }

    if (this.props.isFocus) {
      setTimeout(() => {
        if (this.state.isToggled) {
          const { exchangeSecondaryToPrimaryRatio } = this.props
          if (bns.eq(exchangeSecondaryToPrimaryRatio, '0') || exchangeSecondaryToPrimaryRatio === '') {
            this.toggleCryptoOnBottom()
          } else {
            this.textInputBack && this.textInputBack.focus()
          }
        } else {
          this.textInputFront && this.textInputFront.focus()
        }
      }, 650)
    }
  }

  componentWillUnmount() {
    this.props.flipInputRef(null)
  }

  UNSAFE_componentWillReceiveProps(nextProps: Props) {
    // Check if primary changed first. Don't bother to check secondary if parent passed in a primary
    if (
      nextProps.overridePrimaryDecimalAmount !== this.state.overridePrimaryDecimalAmount ||
      nextProps.forceUpdateGuiCounter !== this.state.forceUpdateGuiCounter
    ) {
      const result = setPrimaryToSecondary(nextProps, sanitizeDecimalAmount(nextProps.overridePrimaryDecimalAmount, nextProps.primaryInfo.maxEntryDecimals))
      this.setState({
        ...result,
        overridePrimaryDecimalAmount: nextProps.overridePrimaryDecimalAmount,
        forceUpdateGuiCounter: nextProps.forceUpdateGuiCounter
      })
    } else {
      // Checks and apply exchange rates
      if (!this.state.isToggled) {
        this.setState(setPrimaryToSecondary(nextProps, this.state.primaryDecimalAmount))
      } else {
        this.setState(setSecondaryToPrimary(nextProps, this.state.secondaryDecimalAmount))
      }
    }
    if (nextProps.primaryInfo.currencyCode !== this.props.primaryInfo.currencyCode) {
      setTimeout(() => this.onKeyPress('0', '', this.props.primaryInfo.maxEntryDecimals, setPrimaryToSecondary), 50)
    }
  }

  // Used on parent ExchangedFlipInput
  toggleCryptoOnBottom = () => {
    if (this.state.isToggled) {
      this.onToggleFlipInput()
    }
  }

  // Used on refs parent (Request Scene)
  textInputBottomFocus = () => {
    if (this.state.isToggled) {
      if (this.textInputBack) {
        this.textInputBack.focus()
      }
    } else {
      if (this.textInputFront) {
        this.textInputFront.focus()
      }
    }
  }

  // Used on refs parent (Request Scene)
  textInputBottomBlur = () => {
    if (this.state.isToggled) {
      if (this.textInputBack) {
        this.textInputBack.blur()
      }
    } else {
      if (this.textInputFront) {
        this.textInputFront.blur()
      }
    }
  }

  onToggleFlipInput = () => {
    this.clipboardMenu.close()
    this.setState({
      isToggled: !this.state.isToggled
    })
    if (this.state.isToggled) {
      if (this.textInputFront) {
        this.textInputFront.focus()
      }
      Animated.spring(this.animatedValue, {
        toValue: 0,
        friction: 8,
        tension: 10,
        useNativeDriver: true
      }).start()
    }
    if (!this.state.isToggled) {
      if (this.textInputBack) {
        this.textInputBack.focus()
      }
      Animated.spring(this.animatedValue, {
        toValue: 1,
        friction: 8,
        tension: 10,
        useNativeDriver: true
      }).start()
    }
  }

  async openClipboardMenu() {
    this.clipboardMenu.close()
    try {
      if ((this.state.textInputFrontFocus || this.state.textInputBackFocus) && (await Clipboard.getString())) {
        this.clipboardMenu.open()
      }
    } catch (error) {
      showError(error)
    }
  }

  handlePasteClipboard = async () => {
    try {
      const clipboard = await Clipboard.getString()
      this.setStateAmounts('', setSecondaryToPrimary)
      if (this.state.isToggled) {
        for (const string of clipboard.split('')) {
          this.onKeyPress(string, this.state.secondaryDecimalAmount, this.props.secondaryInfo.maxEntryDecimals, setSecondaryToPrimary)
        }
      } else {
        for (const string of clipboard.split('')) {
          this.onKeyPress(string, this.state.primaryDecimalAmount, this.props.primaryInfo.maxEntryDecimals, setPrimaryToSecondary)
        }
      }
    } catch (error) {
      showError(error)
    }
  }

  getTextInputFrontRef = (ref: TextInput | null) => {
    this.textInputFront = ref
  }

  textInputFrontFocusTrue = () => {
    this.setState({ textInputFrontFocus: true, rerenderCounter: this.state.rerenderCounter + 1 })
    if (this.props.onFocus) this.props.onFocus()
  }

  textInputFrontFocusFalse = () => {
    this.setState({ textInputFrontFocus: false })
    if (this.props.onBlur) this.props.onBlur()
  }

  textInputFrontFocus = async () => {
    this.openClipboardMenu()
    if (this.textInputFront) {
      this.textInputFront.focus()
    }
  }

  onKeyPress(keyPressed: string, decimalAmount: string, maxEntryDecimals: number, setAmounts: (Props, string) => Amounts) {
    keyPressed = keyPressed.replace(',', '.')
    if (keyPressed === 'Backspace') {
      this.setStateAmounts(truncateDecimals(decimalAmount.slice(0, -1), maxEntryDecimals), setAmounts)
    } else if (checkKeyPress(keyPressed, decimalAmount)) {
      this.setStateAmounts(truncateDecimals(decimalAmount + keyPressed, maxEntryDecimals), setAmounts)
    }
  }

  setStateAmounts(decimalAmount: string, setAmounts: (Props, string) => Amounts) {
    const amounts = setAmounts(this.props, decimalAmount)
    this.setState(amounts, () => this.props.onAmountChanged(amounts.primaryDecimalAmount))
  }

  onPrimaryChangeText = (value: string) => {
    // Hack on android where to only get the number farthest to the right
    if (Platform.OS === 'ios') return
    if (value.match(/[^0-9]/)) return
    this.onKeyPress(value.slice(value.length - 1), this.state.primaryDecimalAmount, this.props.primaryInfo.maxEntryDecimals, setPrimaryToSecondary)
  }

  onPrimaryKeyPress = (e: Event) => {
    if (Platform.OS === 'android' && e.nativeEvent.key.match(/[0-9]/)) return
    this.onKeyPress(e.nativeEvent.key, this.state.primaryDecimalAmount, this.props.primaryInfo.maxEntryDecimals, setPrimaryToSecondary)
  }

  getTextInputBackRef = (ref: TextInput | null) => {
    this.textInputBack = ref
  }

  textInputBackFocusTrue = () => {
    this.setState({ textInputBackFocus: true })
    if (this.props.onFocus) this.props.onFocus()
  }

  textInputBackFocusFalse = () => {
    this.setState({ textInputBackFocus: false })
    if (this.props.onBlur) this.props.onBlur()
  }

  textInputBackFocus = async () => {
    this.openClipboardMenu()
    if (this.textInputBack) {
      this.textInputBack.focus()
    }
  }

  onSecondaryChangeText = (value: string) => {
    // Hack on android where to only get the number farthest to the right
    if (Platform.OS === 'ios') return
    if (value.match(/[^0-9]/)) return
    this.onKeyPress(value.slice(value.length - 1), this.state.secondaryDecimalAmount, this.props.secondaryInfo.maxEntryDecimals, setSecondaryToPrimary)
  }

  onSecondaryKeyPress = (e: Event) => {
    if (Platform.OS === 'android' && e.nativeEvent.key.match(/[0-9]/)) return
    this.onKeyPress(e.nativeEvent.key, this.state.secondaryDecimalAmount, this.props.secondaryInfo.maxEntryDecimals, setSecondaryToPrimary)
  }

  bottomRow = (isFront: boolean) => {
    const { isEditable, inputAccessoryViewID, onNext, topReturnKeyType, theme } = this.props
    const { textInputBackFocus } = this.state
    const styles = getStyles(theme)
    const displayAmount = isFront ? this.state.primaryDisplayAmount : this.state.secondaryDisplayAmount
    const decimalAmount = isFront ? this.state.primaryDecimalAmount : this.state.secondaryDecimalAmount
    const currencyName = isFront ? this.props.primaryInfo.currencyName : this.props.secondaryInfo.currencyName
    const onPress = isFront ? this.textInputFrontFocus : this.textInputBackFocus
    const onChangeText = isFront ? this.onPrimaryChangeText : this.onSecondaryChangeText
    const onKeyPress = isFront ? this.onPrimaryKeyPress : this.onSecondaryKeyPress
    const onFocus = isFront ? this.textInputFrontFocusTrue : this.textInputBackFocusTrue
    const onBlur = isFront ? this.textInputFrontFocusFalse : this.textInputBackFocusFalse
    const ref = isFront ? this.getTextInputFrontRef : this.getTextInputBackRef
    const displayAmountCheck = !decimalAmount || decimalAmount.match(/^0*$/)
    const displayAmountString = displayAmountCheck ? s.strings.string_enter_amount : displayAmount
    const displayAmountStyle = displayAmountCheck ? styles.bottomAmountMuted : styles.bottomAmount

    return (
      <TouchableWithoutFeedback onPress={onPress}>
        <View style={styles.bottomContainer} key="bottom">
          <View style={styles.valueContainer}>
            {(!textInputBackFocus || !displayAmountCheck) && <EdgeText style={displayAmountStyle}>{displayAmountString}</EdgeText>}
            <BlinkingCursor visible={textInputBackFocus} fontSize={1.5} color={theme.deactivatedText} />
          </View>
          {!displayAmountCheck && <EdgeText style={styles.bottomCurrency}>{currencyName}</EdgeText>}
          <TextInput
            style={styles.hiddenTextInput}
            value=""
            onChangeText={onChangeText}
            onKeyPress={onKeyPress}
            autoCorrect={false}
            keyboardType="numeric"
            returnKeyType={topReturnKeyType || 'done'}
            ref={ref}
            onFocus={onFocus}
            onBlur={onBlur}
            editable={isEditable}
            onSubmitEditing={onNext}
            inputAccessoryViewID={inputAccessoryViewID || null}
          />
        </View>
      </TouchableWithoutFeedback>
    )
  }

  topRow = (fieldInfo: FlipInputFieldInfo, amount: string) => {
    const bottomText = `${amount} ${fieldInfo.currencyName}`
    return (
      <TouchableWithoutFeedback onPress={this.onToggleFlipInput} key="bottom">
        <EdgeText>{bottomText}</EdgeText>
      </TouchableWithoutFeedback>
    )
  }

  clipboardRef = (ref: any) => (this.clipboardMenu = ref)

  render() {
    const { primaryInfo, secondaryInfo, headerText, headerLogo, headerCallback, theme } = this.props
    const { isToggled } = this.state
    const frontAnimatedStyle = { transform: [{ rotateX: this.frontInterpolate }] }
    const backAnimatedStyle = { transform: [{ rotateX: this.backInterpolate }] }
    const styles = getStyles(theme)

    return (
      <>
        <TouchableOpacity onPress={headerCallback} style={styles.headerContainer}>
          {headerLogo ? <Image style={styles.headerIcon} source={{ uri: headerLogo }} /> : null}
          {headerCallback ? <RightChevronButton text={headerText} onPress={headerCallback} /> : <EdgeText style={styles.headerText}>{headerText}</EdgeText>}
        </TouchableOpacity>
        <View style={styles.clipboardContainer}>
          <Menu onSelect={this.handlePasteClipboard} ref={this.clipboardRef} renderer={renderers.Popover} rendererProps={{ placement: 'top' }}>
            <MenuTrigger />
            <MenuOptions>
              <MenuOption>
                <EdgeText style={styles.clipboardText}>{s.strings.string_paste}</EdgeText>
              </MenuOption>
            </MenuOptions>
          </Menu>
        </View>
        <View style={styles.flipInputContainer}>
          <View style={styles.flipInput}>
            <Animated.View
              style={[styles.flipInputFront, frontAnimatedStyle, { opacity: this.androidFrontOpacityInterpolate }]}
              pointerEvents={isToggled ? 'none' : 'auto'}
            >
              {this.topRow(secondaryInfo, this.state.secondaryDisplayAmount)}
              {this.bottomRow(true)}
            </Animated.View>
            <Animated.View
              style={[styles.flipInputFront, styles.flipContainerBack, backAnimatedStyle, { opacity: this.androidBackOpacityInterpolate }]}
              pointerEvents={isToggled ? 'auto' : 'none'}
            >
              {this.topRow(primaryInfo, this.state.primaryDisplayAmount)}
              {this.bottomRow(false)}
            </Animated.View>
          </View>
          <ButtonBox onPress={this.onToggleFlipInput} paddingRem={[0.5, 0, 0.5, 1]}>
            <Fontello style={styles.flipIcon} name="exchange" color={theme.iconTappable} size={theme.rem(1.75)} />
          </ButtonBox>
        </View>
      </>
    )
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  // Header
  headerContainer: {
    marginRight: Platform.OS === 'ios' ? theme.rem(3.5) : theme.rem(1.5), // Different because adjustsFontSizeToFit behaves differently on android vs ios
    flexDirection: 'row',
    alignItems: 'center',
    marginBottom: theme.rem(1)
  },
  headerIcon: {
    width: theme.rem(1.5),
    height: theme.rem(1.5),
    marginRight: theme.rem(1)
  },
  headerText: {
    fontWeight: '600',
    fontSize: theme.rem(1.0)
  },

  // Flip Input
  flipInputContainer: {
    flexDirection: 'row',
    alignItems: 'center'
  },
  flipInput: {
    flex: 1,
    paddingRight: theme.rem(0.5)
  },
  flipInputFront: {
    backfaceVisibility: 'hidden'
  },
  flipContainerBack: {
    position: 'absolute',
    top: 0,
    bottom: 0,
    left: 0,
    right: 0
  },
  flipIcon: {
    marginRight: -theme.rem(0.125)
  },

  // Top Amount
  bottomContainer: {
    flexDirection: 'row',
    marginRight: theme.rem(1.5)
  },
  valueContainer: {
    flexDirection: 'row',
    marginRight: theme.rem(0.5)
  },
  bottomAmount: {
    fontFamily: theme.fontFaceMedium,
    fontSize: theme.rem(1.5)
  },
  bottomAmountMuted: {
    fontFamily: theme.fontFaceMedium,
    fontSize: theme.rem(1.5),
    marginLeft: theme.rem(-0.1), // Hack because of amount being bigger font size not aligning to the rest of the text on justified left
    color: theme.deactivatedText
  },
  bottomCurrency: {
    paddingTop: theme.rem(0.125)
  },
  hiddenTextInput: {
    position: 'absolute',
    width: 0,
    height: 0
  },

  // Clipboard Popup
  clipboardContainer: {
    height: 0,
    width: '8%',
    top: theme.rem(0.5),
    alignItems: 'flex-end'
  },
  clipboardText: {
    color: theme.clipboardPopupText,
    padding: theme.rem(0.25)
  }
}))

const FlipInputThemed = withTheme(FlipInputComponent)

// $FlowFixMe - forwardRef is not recognize by flow?
export const FlipInput = React.forwardRef((props, ref) => <FlipInputThemed {...props} flipInputRef={ref} />) // eslint-disable-line
