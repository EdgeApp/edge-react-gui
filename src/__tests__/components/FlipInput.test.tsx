/* globals describe it expect */

import * as React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'

import { getTheme } from '../../components/services/ThemeContext'
import { FlipInputComponent } from '../../components/themed/FlipInput'

describe('FlipInputComponent', () => {
  it('should render with loading props', () => {
    const renderer = new ShallowRenderer()

    const props = {
      overridePrimaryDecimalAmount: 'string',

      // Exchange rate
      exchangeSecondaryToPrimaryRatio: 'string',

      // Information regarding the primary and secondary field. Mostly related to currency name, code, and denominations
      primaryInfo: 'FlipInputFieldInfo',
      secondaryInfo: 'FlipInputFieldInfo',
      onNext: () => undefined,
      onFocus: () => undefined,
      onBlur: () => undefined,
      forceUpdateGuiCounter: 'number',

      // Callback when primaryDecimalAmount changes. **This is only called when the user types into a field or if
      // exchangeSecondaryToPrimaryRatio changes. This does NOT get called when overridePrimaryDecimalAmount is changed by the parent
      onAmountChanged: {
        exchangeAmount: 'string',
        nativeAmount: 'string'
      },
      isEditable: true,
      isFiatOnTop: true,
      isFocus: true,

      topReturnKeyType: 'string',
      inputAccessoryViewID: 'string',
      headerText: 'string',
      headerCallback: () => undefined,
      keyboardVisible: true,
      flipInputRef: () => undefined,
      theme: getTheme()
    }

    const actual = renderer.render(<FlipInputComponent {...props} />)

    expect(actual).toMatchSnapshot()
  })
})
