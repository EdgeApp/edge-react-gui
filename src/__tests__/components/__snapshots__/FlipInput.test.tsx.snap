// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlipInputComponent should render with loading props 1`] = `
<React.Fragment>
  <TouchableOpacity
    onPress={[Function]}
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "marginBottom": 22,
        "marginRight": 79,
      }
    }
  >
    <Memo(CryptoIconComponent)
      marginRem={
        [
          0,
          1,
          0,
          0,
        ]
      }
      sizeRem={1.5}
    />
    <RightChevronButton
      onPress={[Function]}
      text="string"
    />
  </TouchableOpacity>
  <View
    style={
      {
        "alignItems": "flex-end",
        "height": 0,
        "top": 11,
        "width": "8%",
      }
    }
  >
    <ForwardRef(enhanceContext-Context.Consumer(Menu))
      onBackdropPress={[Function]}
      onClose={[Function]}
      onOpen={[Function]}
      onSelect={[Function]}
      renderer={[Function]}
      rendererProps={
        {
          "placement": "top",
        }
      }
    >
      <ForwardRef(enhanceContext-Context.Consumer(MenuTrigger))
        customStyles={{}}
        disabled={false}
      />
      <ForwardRef(enhanceContext-Context.Consumer(MenuOptions))
        customStyles={{}}
      >
        <ForwardRef(enhanceContext-Context.Consumer(MenuOption))
          customStyles={{}}
          disableTouchable={false}
          disabled={false}
        >
          <WithTheme(EdgeTextComponent)
            style={
              {
                "color": "#000000",
                "padding": 6,
              }
            }
          >
            Paste
          </WithTheme(EdgeTextComponent)>
        </ForwardRef(enhanceContext-Context.Consumer(MenuOption))>
      </ForwardRef(enhanceContext-Context.Consumer(MenuOptions))>
    </ForwardRef(enhanceContext-Context.Consumer(Menu))>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        {
          "flex": 1,
          "paddingRight": 11,
        }
      }
    >
      <ForwardRef(AnimatedComponentWrapper)
        pointerEvents="auto"
        style={
          [
            {
              "backfaceVisibility": "hidden",
            },
            {
              "transform": [
                {
                  "rotateX": "0deg",
                },
              ],
            },
            {
              "opacity": 1,
            },
          ]
        }
      >
        <TouchableWithoutFeedback
          onPress={[Function]}
        >
          <WithTheme(EdgeTextComponent)>
            0 undefined
          </WithTheme(EdgeTextComponent)>
        </TouchableWithoutFeedback>
        <TouchableWithoutFeedback
          onPress={[Function]}
        >
          <View
            style={
              {
                "flexDirection": "row",
                "marginRight": 34,
                "minHeight": 45,
              }
            }
          >
            <View
              style={
                {
                  "flexDirection": "row",
                  "marginRight": 11,
                }
              }
            >
              <BlinkingCursor
                showCursor={false}
              />
              <WithTheme(EdgeTextComponent)
                style={
                  {
                    "color": "#87939E",
                    "fontFamily": "Quicksand-Medium",
                    "fontSize": 34,
                    "marginLeft": -2,
                  }
                }
              >
                Amount
              </WithTheme(EdgeTextComponent)>
            </View>
            <WithTheme(EdgeTextComponent)
              style={
                {
                  "color": "#87939E",
                  "paddingTop": 3,
                }
              }
            />
            <TextInput
              autoCorrect={false}
              editable={true}
              inputAccessoryViewID="string"
              keyboardType="numeric"
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              onKeyPress={[Function]}
              onSubmitEditing={[Function]}
              returnKeyType="done"
              style={
                {
                  "height": 0,
                  "position": "absolute",
                  "width": 0,
                }
              }
              value=""
            />
          </View>
        </TouchableWithoutFeedback>
      </ForwardRef(AnimatedComponentWrapper)>
      <ForwardRef(AnimatedComponentWrapper)
        pointerEvents="none"
        style={
          [
            {
              "backfaceVisibility": "hidden",
            },
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            {
              "transform": [
                {
                  "rotateX": "180deg",
                },
              ],
            },
            {
              "opacity": 0,
            },
          ]
        }
      >
        <TouchableWithoutFeedback
          onPress={[Function]}
        >
          <WithTheme(EdgeTextComponent)>
            0 undefined
          </WithTheme(EdgeTextComponent)>
        </TouchableWithoutFeedback>
        <TouchableWithoutFeedback
          onPress={[Function]}
        >
          <View
            style={
              {
                "flexDirection": "row",
                "marginRight": 34,
                "minHeight": 45,
              }
            }
          >
            <View
              style={
                {
                  "flexDirection": "row",
                  "marginRight": 11,
                }
              }
            >
              <BlinkingCursor
                showCursor={false}
              />
              <WithTheme(EdgeTextComponent)
                style={
                  {
                    "color": "#87939E",
                    "fontFamily": "Quicksand-Medium",
                    "fontSize": 34,
                    "marginLeft": -2,
                  }
                }
              >
                Amount
              </WithTheme(EdgeTextComponent)>
            </View>
            <WithTheme(EdgeTextComponent)
              style={
                {
                  "color": "#87939E",
                  "paddingTop": 3,
                }
              }
            />
            <TextInput
              autoCorrect={false}
              editable={true}
              inputAccessoryViewID="string"
              keyboardType="numeric"
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              onKeyPress={[Function]}
              onSubmitEditing={[Function]}
              returnKeyType="done"
              style={
                {
                  "height": 0,
                  "position": "absolute",
                  "width": 0,
                }
              }
              value=""
            />
          </View>
        </TouchableWithoutFeedback>
      </ForwardRef(AnimatedComponentWrapper)>
    </View>
    <ButtonBox
      onPress={[Function]}
      paddingRem={
        [
          0.5,
          0,
          0.5,
          1,
        ]
      }
    >
      <Icon
        allowFontScaling={false}
        color="#00f1a2"
        name="exchange"
        size={39}
        style={
          {
            "marginRight": -3,
          }
        }
      />
    </ButtonBox>
  </View>
</React.Fragment>
`;
